File: <200/122/122.hs
import qualified Data.Map.Strict as M
import qualified Data.Set as S
type Chain     = S.Set Int
type ChainPool = S.Set Chain
type ChainPoolMemo = M.Map Int ChainPool
type PQueue = S.Set (Int, Chain)


main :: IO ()
main = print . sum . map (pred . length . snd) $ representiveChains


representiveChains :: [(Int, [Int])]
representiveChains = M.toList . M.map (S.toList . S.elemAt 0) $ dijkstraLike initMemo initQueue


dijkstraLike :: ChainPoolMemo -> PQueue -> ChainPoolMemo
dijkstraLike memo q
    | S.null q = memo
    | num >= target = dijkstraLike memo popped
    | otherwise = dijkstraLike newMemo newQ
    where
        (num, chainS) = S.elemAt 0 q
        popped  = S.deleteAt 0 q
        chainL  = S.toList chainS
        biggest = maximum chainL
        queries = do
                    a <- chainL
                    b <- filter (\b -> let c = a+b in
                                    -- prune the search tree
                                    a <= b &&
                                    -- prune the search tree
                                    biggest < c &&
                                    -- otherwise no need to search
                                    c <= target &&
                                    -- if the chain = c:chainL is longer than what already found, then it will be discarded by updateQM  
                                    S.notMember (S.fromList (c : chainL)) (memo M.! c)
                                ) chainL
                    let c = a + b
                    let chainS = S.fromList $ c : chainL
                    return (c, chainS)
        -- update the chain memo and the priority queue
        (newMemo, newQ) = foldl updateMQ (memo, popped) queries


updateMQ :: (ChainPoolMemo, PQueue) -> (Int, Chain) -> (ChainPoolMemo, PQueue)
updateMQ (memo, q) (n, chainS)
    | ref  < size = (memo, q)        -- discard (n, set) which is inefficient than what is already found
    | size < ref  = (single, pushed) -- single : the pool is renewed to be S.singleton chainS
    | otherwise   = (multi,  pushed) -- multi  : chainS is added to the pool
    where
        ref    = S.size . S.elemAt 0 $ pool -- length of a chain that is already found 
        pool   = memo M.! n                 -- the pool of chains that are already found
        size   = S.size chainS              -- the size of chain that is newly found
        single = M.update (\_ -> Just . S.singleton $ chainS) n memo
        multi  = M.update (Just . S.insert chainS) n memo
        pushed = S.insert (n, chainS) q


initMemo :: ChainPoolMemo
initMemo = M.fromList $ (1, S.singleton . S.singleton $ 1) : [(n, makeInitPool n) | n <- [2..target]]
    where
        makeInitPool :: Int -> ChainPool
        makeInitPool 1 = S.singleton . S.singleton $ 1
        makeInitPool n = S.singleton . S.fromList $ [1..(n+1)] -- (n+1) ensures the chains never optimal


initQueue :: PQueue
initQueue = S.singleton (1, S.singleton 1)


target :: Int
target = 200

File: <200/122/122.cpp
#include<iostream>
#include<set>
#include<queue>
using namespace std;
#define MAX 200

int shortest_chain_lengths[MAX + 10];
set< set<int> > chain_pool[MAX + 10];
priority_queue< pair<int, set<int> > > q;


void Dijkstra_like(){
  shortest_chain_lengths[1] = 1;
  set<int> tmp = {1};
  q.push({-1, tmp});


  while(q.size()){
    int n = -q.top().first;
    set<int> s = q.top().second;
    q.pop();

    if(s.size() > (*chain_pool[n].begin()).size()){
      continue;
    }
    int l = s.size();

    for(auto itr1 = s.begin(); itr1 != s.end(); ++itr1){
      for(auto itr2 = itr1; itr2 != s.end(); ++itr2){
        int a = *itr1, b = *itr2;
        int c = a + b;

        // no need to search
        if(c > MAX){
          break;
        }

        // restrict the newcomer to be greater than any other so as to prune the tree
        if(c <= (int)*s.rbegin()){
          continue;
        }

        // non-shortest is not required
        if(shortest_chain_lengths[c] < l+1){
          continue;
        }

        // new shortest chain
        if(l+1 < shortest_chain_lengths[c]){
          // update the chain length
          shortest_chain_lengths[c] = l + 1;

          // update the chain pool
          s.insert(c);
          auto singleton = {s};
          chain_pool[c] = singleton;
          
          // update the priority queue
          q.push({-c, s});

          s.erase(c);
        }

        // different path could yield different results for n > c
        else{
          s.insert(c);
          // already-found path is not interesting though. 
          if(chain_pool[c].count(s) == 1){
            s.erase(c);
            continue;
          }

          // update the chain pool
          chain_pool[c].insert(s);

          // update the priority queue
          q.push({-c, s});

          s.erase(c);
        }
      }
    }
  }
}


void preprocess(){
  for(int i = 1; i <= MAX; ++i){
    shortest_chain_lengths[i] = MAX+10;

    set<int> s;
    for(int j = 1; j <= i; ++j){
      s.insert(j);
    }

    chain_pool[i].insert(s);
  }
}


int sum_chain_lengths(){
  int total = 0;
  for(int i = 1; i <= MAX; ++i){
    total += shortest_chain_lengths[i] - 1;
  }

  return total;
}


int main(){
  preprocess();
  Dijkstra_like();
  cout << sum_chain_lengths() << endl;
  return 0;
}